### PR TITLE
Add a `RecordField` that handles `Record<string, DataField>` data

### PR DESCRIPTION
**Pull request by In3luki**
Includes a sample implementation in the battle form rule element.

Validation error:

validation-error

`keys` option set to `["test"]`:

key-error

### CODE REVIEW

**Comment by stwlam** [comment]
We'd want this to extend `ObjectField` and validate the full index signature (keys of `StringField` or `NumberField` and values of whatever).

File: src/module/rules/rule-element/battle-form/rule-element.ts [hunk 1]
```diff
@@ -57,7 +58,19 @@ class BattleFormRuleElement extends RuleElementPF2e<BattleFormRuleSchema> {
                         { required: false, initial: undefined }
                     ),
                     tempHP: new ResolvableValueField({ required: false, nullable: true, initial: null }),
-                    senses: new fields.ObjectField({ required: false, initial: undefined }),
+                    senses: new RecordField(
+                        new fields.StringField({ required: true, blank: false }),
```
Comment (stwlam) [inline src/module/rules/rule-element/battle-form/rule-element.ts:62]: okok but then give this a restricted set of choices: the resulting object should then have an index signature with optional keys

Comment (In3luki) [reply src/module/rules/rule-element/battle-form/rule-element.ts:62]: So no error on choices validation and extract the choices keys as optional keys for the resulting object?

Comment (In3luki) [reply src/module/rules/rule-element/battle-form/rule-element.ts:62]: I think I've got it working now. If the `StringField` props are a specific type those keys are in the resulting object index in addition to allowing any `string` keys.

File: src/module/rules/rule-element/battle-form/schema.ts [hunk 2]
```diff
@@ -27,7 +35,13 @@ type BattleFormRuleOverrideSchema = {
         false
     >;
     tempHP: ResolvableValueField<false, true, true>;
-    senses: ObjectField<BattleFormSenses, BattleFormSenses, false, false, false>;
+    senses: RecordField<
+        StringField<string, string, true, false, false>,
```
Comment (stwlam) [inline src/module/rules/rule-element/battle-form/schema.ts:39]: This should then be the set of of the permitted senses rather than a mere `string`

Comment (In3luki) [reply src/module/rules/rule-element/battle-form/schema.ts]: Done. A normal `Record` in `RecordFieldModelProp` was enough after all.

Comment (In3luki) [reply src/module/rules/rule-element/battle-form/schema.ts]: ok it had to be a partial `Record` after all.

Comment (stwlam) [reply src/module/rules/rule-element/battle-form/schema.ts]: Right, these will globally be `Record<string, something | undefined>`

**Pull request merged**